### PR TITLE
Fix wrong amount in old p ui - use the right amount calculation for payone too

### PR DESCRIPTION
**Pull request by dorin-ciceu**
@plentymarkets/team-order-payment

### CODE REVIEW

File: src/Services/PaymentCreation.php [hunk 1]
```diff
@@ -133,6 +134,26 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
...
         }
 
+        try {
+            /** @var CurrencyExchangeRepositoryContract $currencyService */
+            $currencyService = pluginApp(CurrencyExchangeRepositoryContract::class);
+
+            $defaultCurrency = $currencyService->getDefaultCurrency();
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
```
Comment (estafilarakis) [inline src/Services/PaymentCreation.php]: If `$payment->currency != $defaultCurrency` is true, then we have a foreign currency. In this case the exchange rate should be already set (in the basket or during order creation, if we have an order). Is it necessary to change this value? Maybe you should first check it:
```php
if ($payment->exchangeRatio <= 0.0) {
    $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
}
```

File: src/Services/PaymentCreation.php [hunk 2]
```diff
@@ -235,6 +256,26 @@ public function capturePayment(Payment $payment, Order $order)
...
         $payment->receivedAt = date('Y-m-d H:i:s');
 
+        try {
+            /** @var CurrencyExchangeRepositoryContract $currencyService */
+            $currencyService = pluginApp(CurrencyExchangeRepositoryContract::class);
+
+            $defaultCurrency = $currencyService->getDefaultCurrency();
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
```
Comment (estafilarakis) [inline src/Services/PaymentCreation.php]: same situation

**Review by estafilarakis — COMMENTED** [review]
Why is this necessary? Does PayPal only accept payments in EUR and not in another currency?
Note that the default currency is not always EUR (for British systems it can be GBP and for Swiss systems CHF)!

File: src/Services/PaymentCreation.php (hunk 1 above)
Comment (dorin-ciceu) [reply src/Services/PaymentCreation.php]: I can't find any exchange ratio in the basket. The main problem here is that the payment is saved with the exchange ration = 1 if this code is not added. I also researched this on other payment methods and seems to be the right way.

File: src/Services/PaymentCreation.php [hunk 3]
```diff
@@ -133,6 +134,26 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
...
+                $payment->amount = round(
+                    $currencyService->convertToDefaultCurrency(
+                        $payment->currency,
+                        $payment->amount,
+                        $payment->exchangeRatio
+                    ),
+                    2);
```
Comment (raphael-adam) [inline src/Services/PaymentCreation.php]: This looks wrong to me. I think the original amount in the foreign currency needs to be saved in the payment. Am I right @j-konopka ?

File: src/Services/PaymentCreation.php [hunk 4]
```diff
@@ -133,6 +134,26 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
...
+            }
+        } catch (\Exception $cEx) {
```
Comment (raphael-adam) [inline src/Services/PaymentCreation.php]: Why is this necessary? These empty catch expressions are always a nightmare for debugging.

**Review by raphael-adam — CHANGES_REQUESTED** [review]

File: src/Services/PaymentCreation.php (hunk 3 above)
Comment (j-konopka) [reply src/Services/PaymentCreation.php]: Yes it is better to save the original amount. If their is a foreign currency you have to set the flag "isSystemCurrency" to "false"

Comment (dorin-ciceu) [reply src/Services/PaymentCreation.php]: @j-konopka @raphael-adam ok. We can save the original amount(the amount in foreign currency) but then the old payment UI is not working properly. Also, take a look at the other payment processors, all of them have the same calculation. This bug fix was added because before this fix, the amount was saved in the foreign currency. Note that if we want to save the foreign amount, we have to change it for all payment methods and another problem is that the isSystemCurrency set to false will not work.

File: src/Services/PaymentCreation.php (hunk 4 above)
Comment (dorin-ciceu) [reply src/Services/PaymentCreation.php]: solved

File: src/Services/PaymentCreation.php (hunk 3 above)
Comment (raphael-adam) [reply src/Services/PaymentCreation.php]: Isn't the behaviour like that? :
The amount of the payment is always saved in the currency of the payment. If the payment currency is not the system currency, the exchange rate is saved and `isSystemCurrency` is set to false.
Maybe we can have a call sometime where @j-konopka can explain how the fields should be set.

Comment (j-konopka) [reply src/Services/PaymentCreation.php]: If the `isSystemCurrency` Flag is `true` the amount should be in the system currency. It could be that the currency sign is different to the system currency. But for this case an exchangeRatio should be available.
If the `isSystemCurrency` Flag is `false` the amount should be in the foreign currency and the currency sign should be also the foreign currency.

**Review by estafilarakis — APPROVED** [review]

File: src/Services/PaymentCreation.php [hunk 5]
```diff
@@ -133,6 +134,19 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
...
+
+        /** @var CurrencyExchangeRepositoryContract $currencyService */
+        $currencyService = pluginApp(CurrencyExchangeRepositoryContract::class);
+
+        $defaultCurrency = $currencyService->getDefaultCurrency();
+
+        //when a payment is placed in a foreign currency, we save the foreign amount,
+        // foreign currency sign, exchange ratio and isSystemCurrency set to 0
+        if ($payment->currency != $defaultCurrency) {
+            $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
```
Comment (raphael-adam) [inline src/Services/PaymentCreation.php:146]: I think this is not necessary, when `isSystemCurrency  = false`

File: src/Services/PaymentCreation.php [hunk 6]
```diff
@@ -235,6 +249,26 @@ public function capturePayment(Payment $payment, Order $order)
...
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
+                $payment->amount = round(
+                    $currencyService->convertToDefaultCurrency(
+                        $payment->currency,
+                        $payment->amount,
+                        $payment->exchangeRatio
+                    ),
+                    2);
```
Comment (raphael-adam) [inline src/Services/PaymentCreation.php]: Why is here still the currency convertion? And the code on top looks wrong to me: `$payment->amount = $orderData['amounts'][0]['grossTotal'];`. Maybe lets talk after the daily about this.

File: src/Services/PaymentCreation.php [hunk 7]
```diff
@@ -235,6 +249,26 @@ public function capturePayment(Payment $payment, Order $order)
...
+            }
+        } catch (\Exception $cEx) {
+
+        }
```
Comment (raphael-adam) [inline src/Services/PaymentCreation.php]: Still the empty catch

**Review by raphael-adam — CHANGES_REQUESTED** [review]

**Review by raphael-adam — APPROVED** [review]

**Review by estafilarakis — APPROVED** [review]
